samples/javascript_nodejs/21.corebot-app-insights/bots/dialogAndWelcomeBot.js

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+const { CardFactory } = require('botbuilder');
+const { DialogBot } = require('./dialogBot');
+const WelcomeCard = require('./resources/welcomeCard.json');
+
+class DialogAndWelcomeBot extends DialogBot {
+    /**
+     * @param {ConversationState} conversationState 
+     * @param {UserState} userState 
+     * @param {Dialog} dialog 
+     */
+    constructor(conversationState, userState, dialog) {
+        super(conversationState, userState, dialog);
+
+        this.onMembersAdded(async (turnContext, next) => {
+            const membersAdded = turnContext.activity.membersAdded;
+
+            for (let pos = 0; pos < membersAdded.length; pos++) {
```

Comment:
Couldn't this be done with a `foreach`?

Reply:
Yes, but AFAIK, the `for` loop is more performant than the `foreach` loop ([src](https://stackoverflow.com/a/43032526))